Add getting-started notebooks for azure-quantum with 3rd party Python frameworks

guenp opened the pull request.

cgranade commented on samples/getting-started/azure-quantum/Getting started with Qiskit and IonQ on Azure Quantum.ipynb:
```diff
@@ -0,0 +1,652 @@
+   "id": "ec9beeb2",
+   "metadata": {},
+   "source": [
+    "Below, all cells from the example notebook are copy-pasted. Feel free to give it a try yourself!"
```
We may need a third-party notice for this usage. You should be able to check by adding a cgmanifest.json file to the repo indicating the source for this copy–paste.

guenp replied:
Let's take out this sample completely, I don't think there is much added value anyway.

cgranade replied:
Fair enough! Not arguing for that, though, just so much as making sure we dot our eyes and cross our tees on code reuse.

cgranade commented on samples/getting-started/azure-quantum/Getting started with Qiskit and IonQ on Azure Quantum.ipynb:
```diff
@@ -0,0 +1,652 @@
+    "from qiskit import IBMQ, Aer, transpile, assemble\n",
+    "from qiskit import QuantumCircuit, ClassicalRegister, QuantumRegister\n",
+    "\n",
+    "# import basic plot tools\n",
```
```suggestion
    "# Import basic plotting tools\n",
```

cgranade commented on samples/getting-started/azure-quantum/Getting started with Qiskit and IonQ on Azure Quantum.ipynb:
```diff
@@ -0,0 +1,652 @@
+    "def qft_dagger(qc, n):\n",
+    "    \"\"\"n-qubit QFTdagger the first n qubits in circ\"\"\"\n",
+    "    # Don't forget the Swaps!\n",
+    "    for qubit in range(n//2):\n",
```
It's pretty confusing to refer to the index of a qubit as `qubit`; was that in the original code?

guenp replied:
I ended up taking this code out.

cgranade commented on samples/getting-started/azure-quantum/Getting started with Qiskit and IonQ on Azure Quantum.ipynb:
```diff
@@ -0,0 +1,652 @@
+   "outputs": [],
+   "source": [
+    "def qft_dagger(qc, n):\n",
+    "    \"\"\"n-qubit QFTdagger the first n qubits in circ\"\"\"\n",
```
I'm surprised this didn't flag a warning on cspell?

guenp replied:
I ended up taking this code out.

cgranade commented on samples/getting-started/azure-quantum/Getting started with Qiskit and IonQ on Azure Quantum.ipynb:
```diff
@@ -0,0 +1,652 @@
+    "        qc.swap(qubit, n-qubit-1)\n",
+    "    for j in range(n):\n",
+    "        for m in range(j):\n",
+    "            qc.cp(-math.pi/float(2**(j-m)), m, j)\n",
```
What is the `cp` method here? I'm guessing controlled `P`, corresponding to `Controlled Microsoft.Quantum.Intrinsics.R1`?

guenp replied:
I ended up taking this code out.

cgranade commented on samples/getting-started/azure-quantum/Getting started with Qiskit and IonQ on Azure Quantum.ipynb:
```diff
@@ -0,0 +1,652 @@
+    "    # Don't forget the Swaps!\n",
+    "    for qubit in range(n//2):\n",
+    "        qc.swap(qubit, n-qubit-1)\n",
+    "    for j in range(n):\n",
```
I take it these loop indices were in the original? If not, I'd recommend using more informative loop variable names.

guenp replied:
I ended up taking this code out.

cgranade commented on samples/getting-started/azure-quantum/Getting started with Qiskit and IonQ on Azure Quantum.ipynb:
```diff
@@ -0,0 +1,652 @@
+    "\n",
+    "# Measure of course!\n",
+    "for n in range(3):\n",
+    "    qpe2.measure(n,n)\n",
```
It's odd to use QPE and immediately measure the output of the register, as QPE is used over iterative PE mainly when the output needs to be coherent for use in further quantum processing; as per suggestion above, linking to something about iterative PE may help put this in context, perhaps, as being somewhat a contrived example for the purpose of showing integration?

guenp replied:
I ended up taking this code out.

cgranade commented on samples/getting-started/azure-quantum/Getting started with Qiskit and IonQ on Azure Quantum.ipynb:
```diff
@@ -0,0 +1,652 @@
+   ],
+   "source": [
+    "# Let's see the results!\n",
+    "# aer_sim = Aer.get_backend('aer_simulator')\n",
```
It looks like the call to simulate on Qiskit's noisy simulator is also commented out as well as the `transpile` step? If so, it may be helpful to note that above, too.

guenp replied:
I ended up taking this code out.

cgranade reviewed (commented):
This looks like a good start, but we'll want to make sure this plays well with other places samples can be deployed (e.g.: mybinder) by adding needed README files, dependencies in the Dockerfile, etc.

guenp commented on samples/getting-started/azure-quantum/Getting started with Cirq and Honeywell on Azure Quantum.ipynb:
```diff
@@ -0,0 +1,331 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!pip install azure-quantum[cirq] --quiet"
```
I looked at some of the samples but I don't see any that contain an environment.yml or requirements.txt file. Could you point me to an example?

cgranade replied:
A lot of the samples require only that in the base samples image, and thus don't include a separate environment.yml or requirements.txt, but in cases where something additional is required, such as in the process tomography case (recently moved to https://github.com/microsoft/Quantum/blob/main/samples/characterization/process-tomography/tomography-sample.ipynb, used to be at `samples/interop/python`), we've generally provided a requirements.txt or environment.yml to document that additional requirement.

guenp replied:
I've implemented your suggestion to comment out the `!pip install` lines.

guenp commented on samples/getting-started/azure-quantum/Getting started with Cirq and Honeywell on Azure Quantum.ipynb:
```diff
@@ -0,0 +1,331 @@
+   "source": [
+    "# Getting started with Cirq and Honeywell on Azure Quantum\n",
+    "\n",
+    "This notebooks shows how to send a basic quantum circuit to a Honeywell target via Azure Quantum."
```
To outside users who already have the context of a "cloud service" I think "Azure Quantum" is more clear.

cgranade replied:
Fair, but we've also used that term very explicitly to include not just the cloud service, but also the client SDK and local-only workflows. For instance, the entire docs site is under docs.microsoft.com/azure/quantum, including components unrelated to services.

guenp replied:
That's true but what I meant with context is that for users who are comparing cloud services I imagine they won't append the word "service" behind everything, i.e. they will compare <company name> not <company name> service. This might be something for @bromeg to chime in on.

guenp replied:
@cgranade , I just discussed with @bromeg  and seeing how we are using the term in other parts of the docs she thinks it will be OK to use here and wouldn't create any confusion for the users.

guenp commented on samples/getting-started/azure-quantum/Getting started with Cirq and Honeywell on Azure Quantum.ipynb:
```diff
@@ -0,0 +1,331 @@
+   "id": "5d89f743",
+   "metadata": {},
+   "source": [
+    "You can now run the program via the Azure Quantum service and get the result. The following cell will submit a job that runs the circuit with 100 shots, wait until the job is completed and return the results."
```
Cirq uses the two terms interchangeably and takes an input argument `program` which you use to pass a circuit, so I intentionally used the word `program` here.

guenp replied:
This notebook is meant for Cirq users so I am not sure that kind of explanation is in scope for this notebook.

cgranade replied:
In general, the samples can be seen by anyone browsing the samples repo, our aka.ms/try-qsharp link, or the samples browser. Even if our intent is for the notebook to be read only by Cirq users, that may not be true in practice. If we want to make that a hard requirement for reading the notebook, a comment or note at the top indicating that the notebook assumes familiarity with Cirq along with some links as to where to learn may be helpful. We've done similar elsewhere (probably not enough, to be fair) for other advanced and/or specialized topics, for example, to help users understand the intended audience for each sample.

guenp replied:
Sounds good. I'll add a line with a note on that + a link to the respective documentation.

guenp commented on samples/getting-started/azure-quantum/Getting started with Honeywell on Azure Quantum.ipynb:
```diff
@@ -0,0 +1,283 @@
+{
```
I'll clarify this in the README.

guenp commented on samples/getting-started/azure-quantum/Getting started with Cirq and Honeywell on Azure Quantum.ipynb:
```diff
@@ -0,0 +1,331 @@
+   "metadata": {},
+   "outputs": [],
+   "source": []
+  }
```
Great suggestion - thanks!

cgranade replied:
No worries, happy to help! That's a trick I learned back from the QuTiP folks, they even have it as a policy to always end notebooks that way, going so far as to write a custom magic command to help out with doing so.

guenp commented on samples/getting-started/azure-quantum/Getting started with Honeywell on Azure Quantum.ipynb:
```diff
@@ -0,0 +1,283 @@
+       " <Target name=\"microsoft.substochasticmontecarlo.cpu\", avg. queue time=0 s, Available>,\n",
+       " <Target name=\"ionq.qpu\", avg. queue time=250 s, Available>,\n",
+       " <Target name=\"ionq.simulator\", avg. queue time=0 s, Available>,\n",
+       " <Target name=\"1qbit.tabu\", avg. queue time=0 s, Available>,\n",
+       " <Target name=\"1qbit.pathrelinking\", avg. queue time=0 s, Available>,\n",
+       " <Target name=\"1qbit.pticm\", avg. queue time=0 s, Available>,\n",
+       " <Target name=\"toshiba.sbm.ising\", avg. queue time=0 s, Available>,\n",
```
Not yet, I will change this to filter only on Honeywell targets. Thanks for catching!

guenp commented on samples/getting-started/azure-quantum/Getting started with IonQ on Azure Quantum.ipynb:
```diff
@@ -0,0 +1,403 @@
+    "import pylab as pl\n",
+    "pl.rcParams[\"font.size\"] = 16\n",
+    "hist = {format(n, \"03b\"): 0 for n in range(8)}\n",
+    "hist.update({format(int(k), \"03b\"): v for k, v in results[\"histogram\"].items()})\n",
```
It's typically sparse.

guenp commented on samples/getting-started/azure-quantum/Getting started with Qiskit and Honeywell on Azure Quantum.ipynb:
```diff
@@ -0,0 +1,327 @@
+    "circuit.h(0)\n",
+    "circuit.cx(0, 1)\n",
+    "circuit.cx(1, 2)\n",
+    "circuit.measure([0,1,2], [0, 1, 2])\n",
```
I think documenting how to use cirq is out of scope for this notebook.

cgranade replied:
As with the note on the other sample, I don't think it's documenting how to use Cirq to provide a comment helping users to understand code in our samples.

guenp replied:
Done.

guenp commented on samples/getting-started/azure-quantum/Getting started with Qiskit and IonQ on Azure Quantum.ipynb:
```diff
@@ -0,0 +1,652 @@
+    "\n",
+    "The `azure-quantum[qiskit]` also supports more advanced examples. You can take any existing Qiskit code sample and by changing only a few lines of code, submit it to Azure Quantum.\n",
+    "\n",
+    "Let's use the QPE example from Qiskit textbook: https://qiskit.org/textbook/ch-algorithms/quantum-phase-estimation.html"
```
We don't have a page that explains Quantum Phase Estimation, as far as I recall.

guenp commented on samples/getting-started/azure-quantum/Getting started with Qiskit and IonQ on Azure Quantum.ipynb:
```diff
@@ -0,0 +1,652 @@
+    "import numpy as np\n",
+    "import math\n",
+    "\n",
+    "# importing Qiskit\n",
```
I suggest to keep this as-is as it's copied from an example notebook.

guenp commented on samples/getting-started/azure-quantum/Getting started with Qiskit and Honeywell on Azure Quantum.ipynb:
```diff
@@ -0,0 +1,327 @@
+   ],
+   "source": [
+    "# Create a Quantum Circuit acting on the q register\n",
+    "circuit = QuantumCircuit(3, 3)\n",
```
I think here it's also out of scope for this notebook to explain how the 3rd party library works.

cgranade replied:
Agreed in general, but I mentioned that in this case since it's a bit difficult to read the example without knowing that; the calls to `H` and `CX` are a bit easier to read out of context, I guess?

guenp replied:
I think the notebooks are generally meant for users who are already familiar with Qiskit, so I don't know if this needs to be further clarified.

cgranade replied:
As in the other example, users may discover our samples by any number of different paths. Given that it's not clear reading this code that `circuit` acts on three qubits and three classical bits, I would suggest a comment explaining the code in our sample. Alternatively, if we want to specialize this sample to _require_ Qiskit experience, then we should make that clear in the notebook itself along with links as to further resources.

guenp replied:
I've added a note that we assume some level of Qiskit familiarity, with a link to their docs. Let me know if this addresses your concern, or if you'd recommend a different approach.

cgranade commented on samples/getting-started/azure-quantum/Getting started with IonQ on Azure Quantum.ipynb:
```diff
@@ -0,0 +1,403 @@
+   "outputs": [],
+   "source": [
+    "# See https://docs.ionq.com/#tag/quantum_programs for JSON format spec\n",
+    "circuit = {\n",
```
Perhaps I misread, then; I really wish GitHub's rendering of Notebooks in diffs was not in terms of raw JSON...

guenp replied:
Makes sense, the diffs can be confusing!

guenp commented:
Thanks, @cgranade for your review and suggestions. I believe I've addressed all your comments; I left a few of them as unresolved to make sure we're on the same page.

cgranade commented on binder-index.md:
```diff
@@ -75,6 +75,14 @@ These are noted in the README.md files for each sample, along with complete inst
     </tr>
+    <tr>
+      <td></td>
+      <td><strong><a href="./samples/getting-started/azure-quantum/README.md">Qiskit and Cirq</a></strong></td>
+      <td><a href="./samples/getting-started/azure-quantum">Jupyter notebooks</a></td>
+      <td></td>
+      <td>N/A</td>
```
I'm noticing from this as well that the path is `samples/getting-started/azure-quantum`, which would imply that these notebooks are _the_ way to get started with services? I'd suggest either under `samples/azure-quantum` or `samples/interoperability` would better communicate that this is one of many different ways to work with our services. As a future improvement, we may want to also separate this into multiple folders, following the way other samples are laid out; that will be especially helpful when publishing from the samples repo to other ways of reading samples (e.g.: the Samples Browser).

With respect to the binder index itself, the first column after the sample name has generally been used for quantum programs directly written in a Jupyter Notebook (e.g.: "Q# notebook") rather than programs generated by or called from a host language (e.g.: "Q# + Python"). Following that pattern, using the third column to indicate something like "Cirq/Qiskit + Python" would likely fit with other samples better:

```suggestion
      <td><strong><a href="./samples/getting-started/azure-quantum/README.md">Qiskit and Cirq</a></strong></td>
      <td></td>
      <td><a href="./samples/getting-started/azure-quantum">Cirq/Qiskit + Python</a></td>
      <td>N/A</td>
```

cgranade commented on samples/getting-started/azure-quantum/Getting-started-with-Cirq-and-Honeywell-on-Azure-Quantum.ipynb:
```diff
@@ -0,0 +1,356 @@
+    "\n",
+    "q0, q1 = cirq.LineQubit.range(2)\n",
+    "circuit = cirq.Circuit(\n",
+    "    cirq.H(q0), # Hadamard\n",
```
How in the heck did Hadamard not get added to cspell yet? Wow.

guenp replied:
Oops!

cgranade commented on samples/getting-started/azure-quantum/README.md:
```diff
@@ -0,0 +1,61 @@
+products:
+- qdk
+- azure-quantum
+description: "This sample demonstrates how to use Honeywell and IonQ on the Azure Quantum service with quantum circuits expressed in various Python libraries."
```
This description and the following URL fragment make it sound like using third-party libraries is the only way to use the Azure Quantum service with Python?

guenp replied:
It's more of a statement on how to use the Azure Quantum service with _quantum circuits expressed in Python libraries_.

guenp replied:
I have split up the sample into sub folders so I believe this issue has been addressed.

cgranade commented on samples/getting-started/azure-quantum/README.md:
```diff
@@ -0,0 +1,61 @@
+
+# Getting started with Python libraries on Azure Quantum
+
+These sample notebooks demonstrate how to use the `azure-quantum` Python library to submit quantum circuits expressed with [Cirq](https://quantumai.google/cirq), [Qiskit](https://github.com/QISKit/qiskit-terra) and the provider-specific formats [OpenQASM 2.0](https://github.com/Qiskit/openqasm/tree/OpenQASM2.x) and [JSON](https://docs.ionq.com/#tag/quantum_programs) to an Azure Quantum Workspace.
```
Nit, but if I understand IonQ's documentation correctly, they accept their JSON schema and not JSON in general?

guenp replied:
Correct.

cgranade commented on samples/getting-started/azure-quantum/README.md:
```diff
@@ -0,0 +1,61 @@
+
+In the browser, select the desired sample notebook in your browser to view the code.
+
+## Manifest
```
I think this works for now, but as mentioned above, in the future, we'll likely want to split this up to match other sample layouts.

guenp replied:
I've split up the folder into 3 sub-folders as discussed.

cgranade commented:
Thanks for addressing, @guenp! I think we're almost good to merge this in and defer remaining issues, but it would be good to align on what folder these should be in and how to present these new samples in binder-index.md. Thanks!

cgranade reviewed (commented):
Not sure how my comments got split into two reviews, sorry, but here's the rest.

cgranade reviewed (approved):
LGTM, I'll finish up reorganizing existing samples to make use of the service-oriented samples at getting-started/azure-quantum.
